### PR TITLE
Vortex-Panel Induced Velocity: per-node physics feature from inviscid theory

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Panel Cp (PR #2319) and wake angle (PR #2350) proved that inviscid physics features improve surface pressure prediction. Both encode *surface-only* quantities. This experiment extends the physics oracle to **every mesh node** by computing the 2D velocity field induced by a discrete vortex-panel representation of both foils via the Biot-Savart kernel.

For each mesh node, we sum the induced velocity contributions from all vortex panels:
- u_ind = Σ Γ_i/(2π) · Δy_i/r_i²
- v_ind = Σ -Γ_i/(2π) · Δx_i/r_i²

where Γ_i are vortex strengths from flat-plate theory (Γ ∝ sin(α)), (Δx_i, Δy_i) is the vector from panel i to the query node, and r_i is their distance.

For tandem configurations, both foils contribute separate induced velocity channels: (u_fore, v_fore, u_aft, v_aft). For single-foil, only (u_fore, v_fore) are nonzero (u_aft, v_aft = 0). This gives 4 additional input channels per node.

**Why it helps for tandem:** The aft foil sees a downwash induced by the fore foil's vortex sheet that modifies the effective angle of attack. Panel Cp encodes this only at the surface; induced velocity encodes it at *every* node in the volume, giving the backbone a richer physics context for tandem wake physics. This is the main driver of p_tan error that panel Cp couldn't fully resolve.

**Literature:**
- Hess & Smith (1966) panel method: classical 2D vortex-panel theory
- NeuralFoil (arXiv:2503.16323, Mar 2025): uses panel method as physics oracle
- GeoMPNN (arXiv:2412.09399, NeurIPS 2024): geometry-aware per-node features for mesh learning

## Instructions

Add a `--vortex_panel_velocity` flag to `train.py`. When enabled (along with the existing `--wake_deficit_feature` or independently), compute 4 additional input channels per node.

**Implementation outline:**

1. In the collation / feature-building section, for each sample:
   - Extract surface node positions for fore foil (boundary_id=6) and aft foil (boundary_id=7, tandem-only)
   - Compute N_panels (e.g. 64) control points along each foil surface by subsampling the ordered boundary nodes
   - Compute flat-plate vortex strengths: Γ = (U∞ · sin(AoA)) / N_panels for each panel (uniform distribution)
   - For each mesh node (x_q, y_q), compute induced velocity from each panel (x_p, y_p):
     ```python
     dx = x_q - x_p  # [N_nodes, N_panels]
     dy = y_q - y_p
     r2 = dx**2 + dy**2 + 1e-8
     u_induced = (Gamma / (2 * pi)) * dy / r2  # sum over panels
     v_induced = -(Gamma / (2 * pi)) * dx / r2
     ```
   - Sum over panels: u_fore = u_induced.sum(dim=-1), v_fore = v_induced.sum(dim=-1)
   - For tandem: repeat for aft foil panels → (u_aft, v_aft); zero for single-foil
   - Scale all 4 channels by `--vortex_panel_scale` (default 0.1, same as cp_panel_scale)

2. Concatenate (u_fore, v_fore, u_aft, v_aft) to the per-node input features (adjust input_dim += 4)

3. Add flags:
   - `--vortex_panel_velocity` (bool, enables the feature)
   - `--vortex_panel_scale` (float, default=0.1)
   - `--vortex_panel_n` (int, default=64, number of panels per foil)

4. Run with `--wandb_group round38/vortex-panel` using the full baseline command + `--vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

**Two seeds:** s42 and s73, same as all recent baselines.

**Important:** If the O(N_nodes × N_panels) computation is slow (85K × 64 = 5.4M per foil), use torch broadcasting and ensure it runs in the collation step (not in the training loop). Consider caching the induced velocities in the dataloader if memory allows.

## Baseline

Current best metrics (PR #2350 — Wake Angle Feature, 2-seed avg):

| Metric | Target to beat |
|--------|---------------|
| p_in   | < 11.90       |
| p_oodc | < 7.35        |
| p_tan  | < 27.20       |
| p_re   | < 6.40        |

Reproduce baseline:
```
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature
```

W&B runs: 0gkeylyz (seed 42), sksq5fp5 (seed 73) — entity=wandb-applied-ai-team, project=senpai-v1

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

Starting experiments. Implementation details:

**Approach:**
- Added `compute_vortex_panel_velocity()` function that computes 2D Biot-Savart induced velocity from N=64 discrete vortex panels subsampled from the foil surface nodes
- Flat-plate vortex strength: Γ_i = sin(AoA₀) / N_panels per panel (uniform distribution)
- Channels: (u_fore, v_fore, u_aft, v_aft) × vortex_panel_scale=0.1 = 4 input channels
- Uses AoA₀ for both foils (same as cp_panel convention)
- Panel selection: stride-based subsampling from surface nodes (every k-th node until N=64 panels)
- O(N_nodes × N_panels) = O(85K × 64) = 5.4M per foil, runs in the pre-forward feature-building step

**fun_dim updated:** +4 channels (from 72 → 76)

Running seeds 42 and 73 on GPUs 0 and 1.

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Metrics (2-seed average)

| Metric | Baseline | Seed 42 (`aycq1m8m`) | Seed 73 (`9sk276v6`) | 2-seed avg | Δ vs baseline |
|--------|----------|----------------------|----------------------|------------|---------------|
| **p_in** | **11.90** | 11.755 | 11.988 | **11.872** | -0.2% ✅ |
| **p_oodc** | **7.35** | 7.495 | 7.423 | **7.459** | +1.5% ❌ |
| **p_tan** | **27.20** | 26.221 | 26.417 | **26.319** | -3.2% ✅ |
| **p_re** | **6.40** | 6.178 | 6.280 | **6.229** | -2.7% ✅ |

**3 of 4 metrics beat baseline.** The primary target metric (p_tan) improved by 3.2%.

### Full Surface MAEs (2-seed avg)

| Split | Ux_surf | Uy_surf | p_surf | Ux_vol | Uy_vol | p_vol |
|-------|---------|---------|--------|--------|--------|-------|
| val_in_dist | 2.825 | 0.880 | 11.872 | 0.532 | 0.193 | 12.178 |
| val_ood_cond | 2.004 | 0.633 | 7.459 | 0.347 | 0.144 | 5.875 |
| val_tandem_transfer | 4.083 | 1.625 | 26.319 | 1.286 | 0.593 | 30.576 |
| val_ood_re | 1.782 | 0.447 | 6.229 | 0.504 | 0.268 | 5.220 |

### Commands

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/vortex-panel-s42" --wandb_group "round38/vortex-panel" \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64

# Same for seed 73
```

### W&B Run IDs
- Seed 42: `aycq1m8m` (best epoch 155)
- Seed 73: `9sk276v6` (best epoch 156)

### Peak VRAM
- Seed 42: 40.4 GB
- Seed 73: 39.5 GB

### Analysis

**The hypothesis worked.** Vortex-panel induced velocity improved 3 of 4 metrics, with the largest win on the primary target metric p_tan (-3.2%).

**Why it worked for p_tan:** The Biot-Savart kernel encodes inter-foil downwash at every mesh node, not just the surface. For tandem configurations, the fore foil's circulation induces a non-uniform velocity field in the aft-foil's approach zone. This gives the model volume-wide physics context about the fore→aft coupling, which panel Cp couldn't provide (surface-only). The 3.2% p_tan improvement directly reflects this.

**Why p_oodc regressed (+1.5%):** Extreme-condition samples often have large AoA values (high |sin(AoA)|) which make the induced velocity channels large and variable. The flat-plate vortex strength Γ = sin(AoA)/N_panels overestimates the actual circulation at extreme AoA where thin-airfoil theory breaks down. This may have introduced a noisy signal for OOD-condition samples.

**p_in and p_re improvements:** The induced velocity channels provide useful global flow information even for non-tandem splits. For p_re, the Reynolds number affects circulation but the sin(AoA)/N approximation is scale-free, which might explain the improvement.

**Possible concern:** The per-batch loop over B samples with torch tensors of shape [N, N_panels] adds ~5.4M FP32 operations per foil per sample. In practice, the training throughput is ~5-6 it/s, comparable to baseline (the Biot-Savart computation is dominated by torch broadcast ops and runs efficiently).

### Suggested follow-ups

1. **Chord-length-scaled Γ**: The current Γ = sin(AoA)/N ignores chord length. For the aft foil (smaller chord in some configurations), scaling Γ by the foil chord length would give physically correct circulation. This might further improve p_tan.
2. **Aft-foil AoA for aft panels**: Currently using AoA₀ for both foils. Using AoA₁ (aft foil geometric AoA from x[:, 0, 18]) for aft panels would be physically more accurate.
3. **Investigate p_oodc regression**: The OOD-condition split has extreme AoA/gap/stagger. Could zero out the vortex feature for samples with |AoA| > threshold to avoid noisy features at large AoA.
4. **More panels**: Try N=128 panels — the surface node density might be insufficient at N=64 to capture the full circulation distribution, especially for high-camber foils.
5. **Volume-weighted vortex velocity loss**: Add a small auxiliary loss on the predicted induced velocity to enforce physics consistency.

**Comment by tcapelle** [comment]
ADVISOR: Merging! 3/4 metrics beat baseline: p_tan **-3.2%** (26.319 vs 27.20), p_re **-2.7%** (6.229 vs 6.40), p_in -0.2% (11.872 vs 11.90). Only p_oodc +1.5% regressed — acceptable given the strong tandem and Re-OOD gains. The Biot-Savart induced velocity provides exactly the kind of inter-foil coupling signal we've been looking for: volume-wide physics context about fore→aft wake interaction, which surface-only panel Cp couldn't provide. Your analysis of why p_oodc regressed (thin-airfoil theory breaks down at extreme AoA) is spot on — that's a known limitation of the flat-plate vortex model. Excellent work!
